### DISCUSSION TITLE
New in V4.0.0 - ✨ Improved @editor tool to update your buffers

### DISCUSSION THREAD

**Discussion by olimorris** [olimorris/codecompanion.nvim#260] [Announcements]
Firstly, there's a potential tonne of breaking changes which I've explained in the specific commits. So please consult the [CHANGELOG](https://github.com/olimorris/codecompanion.nvim/blob/main/CHANGELOG.md).

## `@editor` tool

I've made a lot of changes to this tool to try and make it a viable means of updating a buffer from the chat. It's now capable of some pretty complex modifications:

https://github.com/user-attachments/assets/7f0875b4-30d7-4383-a0fb-1e09a9071d54

> Note: I'm using `claude-3-5-sonnet` and it took 454 tokens to make that modification

Previously, the tool could only handle single modifications. Now, it can stack multiple actions together such as `add`, `delete` or `update`. For full disclosure, the XML from the last video was:

```xml
<tools>
  <tool name="editor">
    <action type="delete">
      <buffer>7</buffer>
      <start_line>20</start_line>
      <end_line>28</end_line>
    </action>
    <action type="add">
      <buffer>7</buffer>
      <code>---Count the number of elements in a table (array or non-array)
---@param t table
---@return integer
M.count_improved = function(t)
  if M.is_array(t) then
    return #t
  else
    local count = 0
    for _ in pairs(t) do
      count = count + 1
    end
    return count
  end
end</code>
      <line>20</line>
    </action>
  </tool>
</tools>
```

I've also made sure that the `prompt_library` provides buffer numbers where possible so that you can leverage `@editor` after calling those.

### Call to action

I'd ask the community to try this out and curate the system prompt that comes with the editor tool to make it even better. I've tried the example above with Copilot, Anthropic, OpenAI, Google and all work well. But I expect to hit some edge cases and I'd value your investigation on how to solve them.

## `:CodeCompanionActions` can now use external providers

A valid challenge in recent days was making `:CodeCompanionActions` available for external providers such as Telescope, especially as we reference it as an optional dependency. That is now possible in `v4.0.0` and I'd ask the community to look at that implementation and think about adding [mini.pick](https://github.com/echasnovski/mini.pick) or [fzflua](https://github.com/ibhagwan/fzf-lua).

## Streamlined commands

It's now possible to do `:CodeCompanionChat why is Lua such an awesome language?` to open up a chat buffer and automatically submit the prompt. Saves a few keystrokes and feels more inline with what I would expect if I was using the plugin for the first time.

**Comment by bassamsdata** [comment]
Thank you for all the hard work. I'm testing the new additions.

> I'd ask the community to look at that implementation and consider adding [mini.pick](https://github.com/echasnovski/mini.pick) or [fzf-lua](https://github.com/ibhagwan/fzf-lua).  

I'll work on adding mini.pick in the coming days.

One question, though, it's a bit out of context: what color scheme are you using? I really like the diff colors.

**Reply by olimorris** [reply]
That would be amazing! And I'm using my own [OneDarkPro](https://github.com/olimorris/onedarkpro.nvim) theme. I think it was a contributor who added them actually

**Reply by bassamsdata** [reply]
Thank you! I wasn't aware that you also have a colorscheme.

**Reply by richtong** [reply]
This @editor thing is so great. It does work amazingly well with Sonnet 3.5 and also GPT-4o

I'm happy to look at the @editor prompting, but I don't know the code base well enough to know where it is, but here's testing with other LLMs. BTW your use of XML to do the editing is very clever!

1. Gemini Pro 1.5. It doesn't generate the XML, but does generate correct code, so probably a little more prompting is needed to get it to the emit the XML that you are expecting
2. Llama 3.1-70B Q4 locally on ollama. Doesn't generate the XML
3. Llama 3.2-3B. Doesn't seem to understand the XML desired but is generating code correctly when I look at it. However the code generation locally is amazingly fast and accurate.
4. Deepseek 2.5 is also very good and understands the XML generation

The not surprising ones given they are relatively old are:

1. qwen 2.5 coder. does generate code nor does it understand the XML prompts
2. llama 3.2-90B. This also gets the change in code but not the XML prompts
3. Command-R+. Does the code change but doesn't get how to generate XML


FYI the test I was using was a simple one. Take a parameter that is two calls and make it a global instead and eliminate the parameter

**Reply by olimorris** [reply]
Thanks for testing. Some of the older models are notoriously bad with this type of request unfortunately. I've tried tweaking the system prompts for them but they seemingly struggle with basic XML.  I had really good experience with Gemini Pro. Could you share the specific prompt you sent to it? I'm not actually sure how many people use Gemini as an API endpoint but I'd like to make sure I shake it down properly.

Regarding the code base, just search `editor` and it will come up. I think the repository is in a pretty clean state these days so easy to poke around in. Btw, what's your default LLM of choice?

**Reply by raine** [reply]
Thanks for yet another great release!

What would be the most streamlined way to request changes to the current buffer with minimal keystrokes?

**Reply by S1M0N38** [reply]
I believe that the ability to generate or use XML tags depends on how the model was fine-tuned during the final stages of training. Anthropic's models are among the first to make extensive use of [XML tags for structuring prompts](https://docs.anthropic.com/en/docs/build-with-claude/prompt-engineering/use-xml-tags).

I'm not sure about other models; they might simply be mimicking the usage of XML tags because they are sophisticated enough to do so (just a speculative thought).

**Reply by olimorris** [reply]
I watched a video some weeks ago where a dev was sharing his experiences of working with LLMs in production. The use of tags was something he shared as being key to getting reliable performance from them.

I've made sure the tool prompts have few-shot inference to try and circumvent any lack of training they may have had. Certainly had no issues with the "main" models

**Reply by olimorris** [reply]
@raine I've streamline the system prompt so you can just tag the editor without saying anything
